src/awkward/operations/structure.py

```diff
@@ -2157,6 +2157,202 @@ def getfunction(layout, depth, posaxis):
         return out
 
 
+def _packed(array, axis=None, highlevel=True, behavior=None):
+    layout = ak.operations.convert.to_layout(
+        array, allow_record=False, allow_other=False
+    )
+    nplike = ak.nplike.of(layout)
+
+    def truncate(layout, n):
+        return ak.layout.IndexedArray64(
+            ak.layout.Index64(nplike.arange(n)), layout
+        ).project()
```

Comment:
`truncate` is used in quite a few places, and it would be a lot less expensive if it were implemented as

```python
def truncate(layout, n):
    return layout[:n]
```

That would be zero-copy. Should `ak.packed` copy everything it truncates?

   * If we're truncating a lot, then the original, much larger array could only be freed if we make a copy, and that would be an argument for copying on truncation. However, that's a rare case: they're usually shaving off a little bit, like the roundoff error in RegularArray.
   * In the primary use-case for this function, as a step right before serialization, the copy would be unnecessary because serialization is itself a copy (and it would only take the slice, not the underlying buffer).

So, generally, I think this function should not copy on truncation, returning `layout[:n]` is ideal. But if there's a use-case for copying on truncation, we could call `deep_copy` on all the `Index` attributes and `deep_copy` on `NumpyArray`. (That prevents it from recursing through the node structure, like `project`, but also it uses `memcpy` instead of creating an intermediate array of increasing numbers just to select all the elements in order.)

Reply:
Oh yikes! Good catch - I'd naively [assumed](https://xkcd.com/1339/) it would avoid copies for contiguous counts.